content/courses/advanced-ada/parts/resource_management/controlled_types.rst

```diff
@@ -2343,6 +2343,496 @@ is allocated in the :ada:`Initialize` procedure.
+          Old_Last : constant Positive
+                     := Old_Data'Last;
+          New_Last : constant Positive
+                     := Old_Data'Last + S.Chunk_Size;
+       begin
+          Put_Line ("Increasing Unbounded_Stack "
+                    & "(1 .. "
+                    & Old_Last'Image
+                    & ") to (1 .. "
+                    & New_Last'Image
+                    & ")");
+
+          Reallocate_Data
+            (To         => S.Data,
+             From       => Old_Data,
+             Max_Last   => New_Last,
+             Valid_Last => S.Top);
+
+          Free (Old_Data);
+       end Increase_Size;
+
+       --
+       --  SUBPROGRAMS
+       --
+
+       procedure Push (S : in out Unbounded_Stack;
+                       E :        T) is
+       begin
+          if Is_Full (S) then
+             Increase_Size (S);
+          end if;
+
+          S.Top := S.Top + 1;
+          S.Data (S.Top) := E;
+       end Push;
+
+       function Pop (S : in out Unbounded_Stack)
+                     return T is
+       begin
+          return E : T do
+             if Is_Empty (S) then
+                raise Stack_Underflow;
+             end if;
+
+             E := S.Data (S.Top);
+             S.Top := S.Top - 1;
+          end return;
+       end Pop;
+
+       function Is_Empty (S : Unbounded_Stack)
+                          return Boolean is
+       begin
+          return S.Top = 0;
+       end Is_Empty;
+
+       --
+       --  PRIVATE SUBPROGRAMS
+       --
+
+       procedure Initialize
+         (S : in out Unbounded_Stack)
+       is
+          Last : constant Positive
+                 := S.Chunk_Size;
+       begin
+          Put_Line ("Initializing Unbounded_Stack "
+                    & "(1 .. "
+                    & Last'Image
+                    & ")");
+          S.Data := new T_Array (1 .. S.Chunk_Size);
+       end Initialize;
+
+       procedure Allocate_Duplicate_Data
+         (S : in out Unbounded_Stack)
+       is
+          Last : constant Positive
+                 := S.Data'Last;
+       begin
+          Put_Line ("Duplicating data for new "
+                    & "Unbounded_Stack (1 .. "
+                    & Last'Image
+                    & ")");
+
+          Reallocate_Data
+            (To         => S.Data,
+             From       => S.Data,
+             Max_Last   => Last,
+             Valid_Last => S.Top);
+       end Allocate_Duplicate_Data;
+
+       procedure Adjust
+         (S : in out Unbounded_Stack)
+       is
+       begin
+          Put_Line ("Adjusting Unbounded_Stack...");
+          Allocate_Duplicate_Data (S);
+       end Adjust;
+
+       procedure Finalize
+         (S : in out Unbounded_Stack)
+       is
+          Last : constant Positive
+                 := S.Data'Last;
+       begin
+          Put_Line ("Finalizing Unbounded_Stack "
+                    & "(1 .. "
+                    & Last'Image
+                    & ")");
+          Free (S.Data);
```

Comment:
Finalize might be invoked (automatically) multiple times for the same object S, so the procedure needs to protect itself so that Free is only called if S.Data is not already null.

Reply:
Thanks for the hint! I'll change it.